### ISSUE TITLE
More parameters for option rename function

### ISSUE THREAD

**Issue by ddhp** [sindresorhus/cpy#48]
Hi, first of all, thanks for this amazing library.

I am wondering if it's possible to pass `srcPath` and `dest` to rename function here:
https://github.com/sindresorhus/cpy/blob/master/index.js#L18

What I want to do is to rename the file with the hash value, so I need srcPath to do that.

Thanks!

**Comment by sindresorhus** [comment]
Yeah, I think we should change it from `basename` argument to a context object with useful properties instead:

- `sourceFilename` (previously `basename`)
- `sourcePath`
- `destinationPath`
- What else?

**Comment by sindresorhus** [comment]
Hmm, another idea. We would pass the callback two arguments, `source` and `destination`. Both are a custom file object:

```js
{
	path: '/Users/sindresorhus/foo.txt',
	filename: 'foo.txt',
	fileNameWithoutExtension: 'foo',
	extension: 'txt'
}
```

`source` would be immutable (`Object.freeze()`), while `destination` would be mutable, so to change something, you just set one of the properties. You can set any of the properties and it will be handled correctly. We could use `Proxy` for this.

I think this would make it really nice to use:

```js
const cpy = require('cpy');

(async () => {
	await cpy('foo.js', 'destination', {
		rename: (source, destination) => {
			if (source.filenameWithoutExtension === 'foo') {
				destination.filenameWithoutExtension = 'bar';
			}
		}
	});
})();
```

If you just want to change the extension, you could do:

```js
const cpy = require('cpy');

(async () => {
	await cpy('foo.js', 'destination', {
		rename: (source, destination) => {
			if (source.filenameWithoutExtension === 'foo') {
				destination.extension = 'ts';
			}

			console.log(destination.filename);
			//=> 'foo.ts'
		}
	});
})();
```

**Comment by adamduncan** [comment]
@sindresorhus This does sound like a nice approach indeed. Was there desire to move ahead with this?

**Comment by sindresorhus** [comment]
My above proposal is how it should be implemented. It's not something I plan to work on anytime soon though. However, good PRs are always welcome.

**Comment by davidliudev** [comment]
A good feature is if there is a name clash then rename the file.
need this mechanism for it to work.
